File: jvm/periph-java/src/main/java/it/uhde/periph/chips/environmental/Bme280Minimal.java
package it.uhde.periph.chips.environmental;

import it.uhde.periph.transport.Transport;

import java.io.IOException;

/**
 * BME280 — combined humidity + pressure + temperature sensor (minimal driver).
 *
 * <p>Reads calibrated temperature (°C), pressure (hPa), and humidity (%RH) via
 * I²C. The chip ID register is verified to be {@code 0x60}.
 *
 * <p>Sibling of the BMP280 driver: register-compatible for pressure and
 * temperature, plus an integrated humidity front-end (its own calibration
 * block, control register, output registers, and compensation formula).
 *
 * <p>Default settings: osrs_t=×1, osrs_p=×1, osrs_h=×1, IIR filter off,
 * forced mode.
 */
public class Bme280Minimal {

    // Register addresses
    protected static final int REG_CALIB      = 0x88;
    protected static final int REG_H1         = 0xA1;
    protected static final int REG_ID         = 0xD0;
    protected static final int REG_SOFT_RST   = 0xE0;
    protected static final int REG_CAL_H2     = 0xE1;
    protected static final int REG_CTRL_HUM   = 0xF2;
    protected static final int REG_STATUS     = 0xF3;
    protected static final int REG_CTRL_MEAS  = 0xF4;
    protected static final int REG_CONFIG     = 0xF5;
    protected static final int REG_DATA       = 0xF7;

    protected static final int CHIP_ID      = 0x60;
    protected static final int RESET_CMD    = 0xB6;
    protected static final int MEAS_TIME_MS = 9;

    protected final Transport transport;

    // Calibration coefficients
    protected int digT1;   // uint16
    protected int digT2;   // int16
    protected int digT3;   // int16
    protected int digP1;   // uint16
    protected int digP2;   // int16
    protected int digP3;   // int16
    protected int digP4;   // int16
    protected int digP5;   // int16
    protected int digP6;   // int16
    protected int digP7;   // int16
    protected int digP8;   // int16
    protected int digP9;   // int16
    protected int digH1;   // uint8
    protected int digH2;   // int16
    protected int digH3;   // uint8
    protected int digH4;   // int16 (12-bit sign-extended)
    protected int digH5;   // int16 (12-bit sign-extended)
    protected int digH6;   // int8

    /** tFine shared between temperature, pressure, and humidity compensation. */
    protected int tFine;

    /** ctrl_hum value applied at each measurement. */
    protected int ctrlHum  = 0x01;
    /** ctrl_meas value applied at each measurement. */
    protected int ctrlMeas = 0x25;
    /** config register value. */
    protected int config   = 0x00;

    /**
     * Construct the driver, verify the chip ID, and load calibration data.
     *
     * @param transport I²C transport bound to the BME280 address (0x76 or 0x77)
     * @throws IOException on I²C error, wrong chip ID, or invalid calibration
     */
    public Bme280Minimal(Transport transport) throws IOException {
        this(transport, 0x76);
    }

    /**
     * Construct the driver at the given address, verify the chip ID, and load
     * calibration data.
     *
     * @param transport I²C transport bound to the given address
     * @param addr      I²C device address (0x76 or 0x77)
     * @throws IOException on I²C error, wrong chip ID, or invalid calibration
     */
    public Bme280Minimal(Transport transport, int addr) throws IOException {
        this.transport = transport;

        byte[] id = transport.writeRead(new byte[]{(byte) REG_ID}, 1);
        int chipId = id[0] & 0xFF;
        if (chipId != CHIP_ID) {
            throw new IOException(
                    "BME280 not found: expected 0x60, got 0x"
                    + Integer.toHexString(chipId));
        }

        readCalibration();

        transport.write(new byte[]{(byte) REG_CTRL_HUM, (byte) ctrlHum});
        transport.write(new byte[]{(byte) REG_CTRL_MEAS, (byte) ctrlMeas});
        transport.write(new byte[]{(byte) REG_CONFIG, (byte) config});
    }

    /**
     * Read and unpack the 33-byte calibration block from NVM (26 bytes from
     * 0x88 plus 7 bytes from 0xE1).
     *
     * <p>Calibration is little-endian for the 16-bit fields. The humidity
     * block 0xE1–0xE7 packs dig_H4 / dig_H5 into register 0xE5 (lower / upper
     * nibble); both are 12-bit signed values sign-extended to 16-bit.
     *
     * @throws IOException on I²C error
     */
    protected void readCalibration() throws IOException {
        byte[] cal = transport.writeRead(new byte[]{(byte) REG_CALIB}, 26);
        digT1 = ((cal[1] & 0xFF) << 8) | (cal[0] & 0xFF);
        digT2 = (short) (((cal[3] & 0xFF) << 8) | (cal[2] & 0xFF));
        digT3 = (short) (((cal[5] & 0xFF) << 8) | (cal[4] & 0xFF));
        digP1 = ((cal[7] & 0xFF) << 8) | (cal[6] & 0xFF);
        digP2 = (short) (((cal[9] & 0xFF) << 8) | (cal[8] & 0xFF));
        digP3 = (short) (((cal[11] & 0xFF) << 8) | (cal[10] & 0xFF));
        digP4 = (short) (((cal[13] & 0xFF) << 8) | (cal[12] & 0xFF));
        digP5 = (short) (((cal[15] & 0xFF) << 8) | (cal[14] & 0xFF));
        digP6 = (short) (((cal[17] & 0xFF) << 8) | (cal[16] & 0xFF));
        digP7 = (short) (((cal[19] & 0xFF) << 8) | (cal[18] & 0xFF));
        digP8 = (short) (((cal[21] & 0xFF) << 8) | (cal[20] & 0xFF));
        digP9 = (short) (((cal[23] & 0xFF) << 8) | (cal[22] & 0xFF));
        digH1 = cal[25] & 0xFF;

        byte[] h = transport.writeRead(new byte[]{(byte) REG_CAL_H2}, 7);
        digH2 = (short) (((h[1] & 0xFF) << 8) | (h[0] & 0xFF));
        digH3 = h[2] & 0xFF;
        int h4raw = ((h[3] & 0xFF) << 4) | (h[4] & 0x0F);
        int h5raw = ((h[5] & 0xFF) << 4) | ((h[4] >> 4) & 0x0F);
        digH4 = (h4raw & 0x800) != 0 ? (short) (h4raw | 0xF000) : (short) h4raw;
        digH5 = (h5raw & 0x800) != 0 ? (short) (h5raw | 0xF000) : (short) h5raw;
        digH6 = (byte) h[6];
    }

    /**
     * Trigger a forced-mode measurement and burst-read 8 bytes from 0xF7.
     *
     * <p>Writes ctrl_hum (must precede ctrl_meas for humidity oversampling to
     * latch), then writes ctrl_meas with forced mode, waits 9 ms, and reads
     * press[19:0], temp[19:0], and hum[15:0] in a single burst.
     *
     * @return 8-byte raw data array: [press_msb, press_lsb, press_xlsb,
     *         temp_msb, temp_lsb, temp_xlsb, hum_msb, hum_lsb]
     * @throws IOException on I²C error
     */
    protected byte[] triggerAndRead() throws IOException {
        transport.write(new byte[]{(byte) REG_CTRL_HUM, (byte) ctrlHum});
        transport.write(new byte[]{(byte) REG_CTRL_MEAS, (byte) ((ctrlMeas & 0xFC) | 0x01)});
        try { Thread.sleep(MEAS_TIME_MS); } catch (InterruptedException e) { Thread.currentThread().interrupt(); }
        return transport.writeRead(new byte[]{(byte) REG_DATA}, 8);
    }

    /**
     * Compute temperature compensation and update tFine.
     *
     * @param adcT raw 20-bit temperature ADC value
     * @return temperature in °C
     */
    protected double compensateTemperature(int adcT) {
        long var1 = (((long) adcT >> 3) - ((long) digT1 << 1)) * (long) digT2 >> 11;
        long var2 = ((((long) adcT >> 4) - (long) (digT1 & 0xFFFFL))
                   * (((long) adcT >> 4) - (long) (digT1 & 0xFFFFL)) >> 12)
                   * (long) digT3 >> 14;
        tFine = (int) (var1 + var2);
        return ((tFine * 5 + 128) >> 8) / 100.0;
    }

    /**
     * Compute pressure compensation using the current tFine value.
     *
     * @param adcP raw 20-bit pressure ADC value
     * @return pressure in hPa
     */
    protected double compensatePressure(int adcP) {
        long var1 = (long) tFine - 128000L;
        long var2 = var1 * var1 * (long) digP6;
        var2 = var2 + ((var1 * (long) digP5) << 17);
        var2 = var2 + ((long) digP4 << 35);
        var1 = ((var1 * var1 * (long) digP3) >> 8) + ((var1 * (long) digP2) << 12);
        var1 = ((1L << 47) + var1) * ((long) digP1 & 0xFFFFL) >> 33;
        if (var1 == 0) return 0.0;
        long p = 1048576L - (long) adcP;
        p = ((p << 31) - var2) * 3125L / var1;
        var1 = ((long) digP9 * (p >> 13) * (p >> 13)) >> 25;
        var2 = ((long) digP8 * p) >> 19;
        p = ((p + var1 + var2) >> 8) + ((long) digP7 << 4);
        return (p / 256.0) / 100.0;
    }

    /**
     * Compute humidity compensation using the current tFine value.
     *
     * @param adcH raw 16-bit humidity ADC value
     * @return humidity in %RH
     */
    protected double compensateHumidity(int adcH) {
        long v = (long) tFine - 76800L;
        long vX1 = (((long) adcH << 14) - ((long) digH4 << 20) - ((long) digH5 * v) + 16384) >> 15;
        v = ((v * (long) digH6) >> 10) * (((v * (long) digH3) >> 11) + 32768L);
        v = v >> 10;
        v = v + 2097152L;
        v = ((v * (long) digH2) + 8192L) >> 14;
        v = vX1 * v;
        long vX2 = (v >> 15) * (v >> 15);
        v = v - (((vX2 >> 7) * (long) digH1) >> 4);
        if (v < 0) v = 0;
        if (v > 419430400L) v = 419430400L;
        return (v >> 12) / 1024.0;
    }

    /**
     * Read the temperature.
     *
     * <p>Triggers a forced-mode measurement and returns the result in degrees
     * Celsius. Also updates tFine for subsequent pressure/humidity reads.
     *
     * @return temperature in °C
     * @throws IOException on I²C error
     */
    public double temperature() throws IOException {
        byte[] raw = triggerAndRead();
        int adcT = ((raw[3] & 0xFF) << 12) | ((raw[4] & 0xFF) << 4) | ((raw[5] & 0xFF) >> 4);
        return compensateTemperature(adcT);
    }

    /**
     * Read the pressure.
     *
     * <p>Triggers a forced-mode measurement, compensates temperature first (to
     * populate tFine), then compensates pressure. Returns the result in hPa.
     *
     * @return pressure in hPa
     * @throws IOException on I²C error
     */
    public double pressure() throws IOException {
        byte[] raw = triggerAndRead();
        int adcP = ((raw[0] & 0xFF) << 12) | ((raw[1] & 0xFF) << 4) | ((raw[2] & 0xFF) >> 4);
        int adcT = ((raw[3] & 0xFF) << 12) | ((raw[4] & 0xFF) << 4) | ((raw[5] & 0xFF) >> 4);
        compensateTemperature(adcT);
        return compensatePressure(adcP);
    }

    /**
     * Read the humidity.
     *
     * <p>Triggers a forced-mode measurement, compensates temperature first (to
     * populate tFine), then compensates humidity. Returns the result in %RH.
     *
     * @return relative humidity in %RH
     * @throws IOException on I²C error
     */
    public double humidity() throws IOException {
        byte[] raw = triggerAndRead();
        int adcH = ((raw[6] & 0xFF) << 8) | (raw[7] & 0xFF);
        int adcT = ((raw[3] & 0xFF) << 12) | ((raw[4] & 0xFF) << 4) | ((raw[5] & 0xFF) >> 4);
        compensateTemperature(adcT);
        return compensateHumidity(adcH);
    }
}

File: jvm/periph-kotlin/src/main/kotlin/it/uhde/periph/chips/environmental/Bme280Minimal.kt
package it.uhde.periph.chips.environmental

import it.uhde.periph.transport.Transport
import java.io.IOException

/**
 * BME280 — combined humidity + pressure + temperature sensor (minimal driver).
 *
 * Reads calibrated temperature (°C), pressure (hPa), and humidity (%RH) via
 * I²C. The chip ID register is verified to be 0x60.
 *
 * Sibling of the BMP280 driver: register-compatible for pressure and
 * temperature, plus an integrated humidity front-end (its own calibration
 * block, control register, output registers, and compensation formula).
 *
 * Default settings: osrs_t=×1, osrs_p=×1, osrs_h=×1, IIR filter off,
 * forced mode.
 *
 * Configurable I²C address: 0x76 (SDO low, default) or 0x77 (SDO high).
 */
open class Bme280Minimal @JvmOverloads constructor(
    protected val transport: Transport,
    addr: Int = 0x76
) {

    companion object {
        const val REG_CALIB       = 0x88
        const val REG_H1          = 0xA1
        const val REG_ID          = 0xD0
        const val REG_SOFT_RST    = 0xE0
        const val REG_CAL_H2      = 0xE1
        const val REG_CTRL_HUM    = 0xF2
        const val REG_STATUS      = 0xF3
        const val REG_CTRL_MEAS   = 0xF4
        const val REG_CONFIG      = 0xF5
        const val REG_DATA        = 0xF7

        const val CHIP_ID         = 0x60
        const val RESET_CMD       = 0xB6
        const val MEAS_TIME_MS    = 9L
    }

    protected var digT1: Int = 0
    protected var digT2: Int = 0
    protected var digT3: Int = 0
    protected var digP1: Int = 0
    protected var digP2: Int = 0
    protected var digP3: Int = 0
    protected var digP4: Int = 0
    protected var digP5: Int = 0
    protected var digP6: Int = 0
    protected var digP7: Int = 0
    protected var digP8: Int = 0
    protected var digP9: Int = 0
    protected var digH1: Int = 0
    protected var digH2: Int = 0
    protected var digH3: Int = 0
    protected var digH4: Int = 0
    protected var digH5: Int = 0
    protected var digH6: Int = 0

    protected var tFine: Int = 0

    protected var ctrlHum: Int = 0x01
    protected var ctrlMeas: Int = 0x25
    protected var config: Int = 0x00

    init {
        val id = transport.writeRead(byteArrayOf(REG_ID.toByte()), 1)
        val chipId = id[0].toInt() and 0xFF
        if (chipId != CHIP_ID) {
            throw IOException("BME280 not found: expected 0x60, got 0x" + Integer.toHexString(chipId))
        }

        readCalibration()

        transport.write(byteArrayOf(REG_CTRL_HUM.toByte(), ctrlHum.toByte()))
        transport.write(byteArrayOf(REG_CTRL_MEAS.toByte(), ctrlMeas.toByte()))
        transport.write(byteArrayOf(REG_CONFIG.toByte(), config.toByte()))
    }

    /**
     * Read and unpack the 33-byte calibration block from NVM (26 bytes from
     * 0x88 plus 7 bytes from 0xE1).
     *
     * Calibration is little-endian for the 16-bit fields. The humidity block
     * 0xE1–0xE7 packs dig_H4 / dig_H5 into register 0xE5 (lower / upper
     * nibble); both are 12-bit signed values sign-extended to 16-bit.
     *
     * @throws IOException on I²C error
     */
    protected fun readCalibration() {
        val cal = transport.writeRead(byteArrayOf(REG_CALIB.toByte()), 26)
        digT1 = ((cal[1].toInt() and 0xFF) shl 8) or (cal[0].toInt() and 0xFF)
        digT2 = ((cal[3].toInt() and 0xFF) shl 8) or (cal[2].toInt() and 0xFF)
        digT3 = ((cal[5].toInt() and 0xFF) shl 8) or (cal[4].toInt() and 0xFF)
        digP1 = ((cal[7].toInt() and 0xFF) shl 8) or (cal[6].toInt() and 0xFF)
        digP2 = ((cal[9].toInt() and 0xFF) shl 8) or (cal[8].toInt() and 0xFF)
        digP3 = ((cal[11].toInt() and 0xFF) shl 8) or (cal[10].toInt() and 0xFF)
        digP4 = ((cal[13].toInt() and 0xFF) shl 8) or (cal[12].toInt() and 0xFF)
        digP5 = ((cal[15].toInt() and 0xFF) shl 8) or (cal[14].toInt() and 0xFF)
        digP6 = ((cal[17].toInt() and 0xFF) shl 8) or (cal[16].toInt() and 0xFF)
        digP7 = ((cal[19].toInt() and 0xFF) shl 8) or (cal[18].toInt() and 0xFF)
        digP8 = ((cal[21].toInt() and 0xFF) shl 8) or (cal[20].toInt() and 0xFF)
        digP9 = ((cal[23].toInt() and 0xFF) shl 8) or (cal[22].toInt() and 0xFF)
        digH1 = cal[25].toInt() and 0xFF

        val h = transport.writeRead(byteArrayOf(REG_CAL_H2.toByte()), 7)
        digH2 = ((h[1].toInt() and 0xFF) shl 8) or (h[0].toInt() and 0xFF)
        digH3 = h[2].toInt() and 0xFF
        val h4raw = ((h[3].toInt() and 0xFF) shl 4) or (h[4].toInt() and 0x0F)
        val h5raw = ((h[5].toInt() and 0xFF) shl 4) or ((h[4].toInt() shr 4) and 0x0F)
        digH4 = if ((h4raw and 0x800) != 0) (h4raw or 0xF000).toShort().toInt() else h4raw.toShort().toInt()
        digH5 = if ((h5raw and 0x800) != 0) (h5raw or 0xF000).toShort().toInt() else h5raw.toShort().toInt()
        digH6 = h[6].toByte().toInt()
    }

    /**
     * Trigger a forced-mode measurement and burst-read 8 bytes from 0xF7.
     *
     * Writes ctrl_hum (must precede ctrl_meas for humidity oversampling to
     * latch), then writes ctrl_meas with forced mode, waits 9 ms, and reads
     * press[19:0], temp[19:0], and hum[15:0] in a single burst.
     *
     * @return 8-byte raw data array
     * @throws IOException on I²C error
     */
    protected fun triggerAndRead(): ByteArray {
        transport.write(byteArrayOf(REG_CTRL_HUM.toByte(), ctrlHum.toByte()))
        transport.write(byteArrayOf(REG_CTRL_MEAS.toByte(), ((ctrlMeas and 0xFC) or 0x01).toByte()))
        try { Thread.sleep(MEAS_TIME_MS) } catch (e: InterruptedException) { Thread.currentThread().interrupt() }
        return transport.writeRead(byteArrayOf(REG_DATA.toByte()), 8)
    }

    /**
     * Compute temperature compensation and update tFine.
     *
     * @param adcT raw 20-bit temperature ADC value
     * @return temperature in °C
     */
    protected fun compensateTemperature(adcT: Int): Double {
        val var1 = (((adcT.toLong() shr 3) - (digT1.toLong() shl 1)) * digT2.toLong()) shr 11
        val var2 = ((((adcT.toLong() shr 4) - (digT1.toLong() and 0xFFFFL)) *
                    ((adcT.toLong() shr 4) - (digT1.toLong() and 0xFFFFL))) shr 12) * digT3.toLong() shr 14
        tFine = (var1 + var2).toInt()
        return ((tFine * 5 + 128) shr 8) / 100.0
    }

    /**
     * Compute pressure compensation using the current tFine value.
     *
     * @param adcP raw 20-bit pressure ADC value
     * @return pressure in hPa
     */
    protected fun compensatePressure(adcP: Int): Double {
        val t = tFine.toLong()
        var var1 = t - 128000L
        var var2 = var1 * var1 * digP6.toLong()
        var2 += (var1 * digP5.toLong()) shl 17
        var2 += digP4.toLong() shl 35
        var1 = ((var1 * var1 * digP3.toLong()) shr 8) + ((var1 * digP2.toLong()) shl 12)
        var1 = ((1L shl 47) + var1) * (digP1.toLong() and 0xFFFFL) shr 33
        if (var1 == 0L) return 0.0
        var p = 1048576L - adcP.toLong()
        p = ((p shl 31) - var2) * 3125L / var1
        var1 = (digP9.toLong() * (p shr 13) * (p shr 13)) shr 25
        var2 = (digP8.toLong() * p) shr 19
        p = ((p + var1 + var2) shr 8) + (digP7.toLong() shl 4)
        return (p / 256.0) / 100.0
    }

    /**
     * Compute humidity compensation using the current tFine value.
     *
     * @param adcH raw 16-bit humidity ADC value
     * @return humidity in %RH
     */
    protected fun compensateHumidity(adcH: Int): Double {
        var v = tFine.toLong() - 76800L
        val vX1 = (((adcH.toLong() shl 14) - (digH4.toLong() shl 20) - (digH5.toLong() * v) + 16384) shr 15)
        v = ((v * digH6.toLong()) shr 10) * (((v * digH3.toLong()) shr 11) + 32768L)
        v = v shr 10
        v += 2097152L
        v = ((v * digH2.toLong()) + 8192L) shr 14
        v = vX1 * v
        val vX2 = (v shr 15) * (v shr 15)
        v -= (((vX2 shr 7) * digH1.toLong()) shr 4)
        if (v < 0) v = 0
        if (v > 419430400L) v = 419430400L
        return (v shr 12) / 1024.0
    }

    /**
     * Read the temperature.
     *
     * Triggers a forced-mode measurement and returns the result in degrees
     * Celsius. Also updates tFine for subsequent pressure/humidity reads.
     *
     * @return temperature in °C
     * @throws IOException on I²C error
     */
    fun temperature(): Double {
        val raw = triggerAndRead()
        val adcT = ((raw[3].toInt() and 0xFF) shl 12) or ((raw[4].toInt() and 0xFF) shl 4) or ((raw[5].toInt() and 0xFF) shr 4)
        return compensateTemperature(adcT)
    }

    /**
     * Read the pressure.
     *
     * Triggers a forced-mode measurement, compensates temperature first (to
     * populate tFine), then compensates pressure. Returns the result in hPa.
     *
     * @return pressure in hPa
     * @throws IOException on I²C error
     */
    fun pressure(): Double {
        val raw = triggerAndRead()
        val adcP = ((raw[0].toInt() and 0xFF) shl 12) or ((raw[1].toInt() and 0xFF) shl 4) or ((raw[2].toInt() and 0xFF) shr 4)
        val adcT = ((raw[3].toInt() and 0xFF) shl 12) or ((raw[4].toInt() and 0xFF) shl 4) or ((raw[5].toInt() and 0xFF) shr 4)
        compensateTemperature(adcT)
        return compensatePressure(adcP)
    }

    /**
     * Read the humidity.
     *
     * Triggers a forced-mode measurement, compensates temperature first (to
     * populate tFine), then compensates humidity. Returns the result in %RH.
     *
     * @return relative humidity in %RH
     * @throws IOException on I²C error
     */
    fun humidity(): Double {
        val raw = triggerAndRead()
        val adcH = ((raw[6].toInt() and 0xFF) shl 8) or (raw[7].toInt() and 0xFF)
        val adcT = ((raw[3].toInt() and 0xFF) shl 12) or ((raw[4].toInt() and 0xFF) shl 4) or ((raw[5].toInt() and 0xFF) shr 4)
        compensateTemperature(adcT)
        return compensateHumidity(adcH)
    }
}
